### PR TITLE
Implement Direct Message Chat

### PR DESCRIPTION
**Pull request by MahemaS27**
- Added logic in `recieveMessage` method in controller to handle a `DirectMessage` case
- Added helper method to find correct listener to send message to
- Updated commenting for methods in `CoveyTownController`
- Added tests for new functionality
- New method in `TestUtils` for testing DMs for socket

### CODE REVIEW

**Comment by abigailhodge** [comment]
Please add a test for the new Socket logic in `CoveyTownsSocket.test.ts`

File: services/roomService/src/client/CoveyTownsSocket.test.ts [hunk 1]
```diff
@@ -54,119 +60,274 @@ describe('TownServiceApiSocket', () => {
...
+    );
+
+    const message = TestUtils.createDirectMessageForTesting(
+      joinData2.currentPlayers[0]._id,
+      joinData2.currentPlayers[1]._id,
+    );
+    // emit the direct message action from socket
+    socket.emit('messageSent', message);
+    // message should be received by the second player
+    const [receivedMessage] = await Promise.all([messageReceived]);
```
Comment (abigailhodge) [inline services/roomService/src/client/CoveyTownsSocket.test.ts]: I think you should make sure that it goes to both players: i.e. when you're making joinData2, i.e. create two socket clients based on the two `apiClient.joinTown` data, set up `messageReceived` Promises for both of them, and make sure they both resolve/that the received messages both match `message`

File: services/roomService/src/client/TestUtils.ts [hunk 2]
```diff
@@ -123,6 +127,17 @@ export function createMessageForTesting(type: MessageType, player1: Player): Mes
       directMessageId: null,
     };
   }
+  if (type === MessageType.DirectMessage && player2) {
+    return {
+      userName: nanoid(),
+      userId: player1.id,
+      location: player1.location,
```
Comment (abigailhodge) [inline services/roomService/src/client/TestUtils.ts]: Is this necessary with the `createDirectMessageForTesting` test?

File: services/roomService/src/lib/CoveyTownController.test.ts [hunk 3]
```diff
@@ -158,7 +157,33 @@ describe('CoveyTownController', () => {
...
+      // create a direct message between player 1 and player 2
+      const message = TestUtils.createMessageForTesting(
+        MessageType.DirectMessage,
+        player1,
+        player2,
+      );
+      testingTown.receiveMessage(message);
+
+      // check that player 2 recieved the direct message
+      expect(directMockListeners[0].onMessageReceived).toBeCalledWith(message);
```
Comment (abigailhodge) [inline services/roomService/src/lib/CoveyTownController.test.ts:185]: Again, I think you should create a listener for player1 and make sure that they're also receiving their own message

File: services/roomService/src/lib/CoveyTownController.ts [hunk 4]
```diff
@@ -195,4 +207,17 @@ export default class CoveyTownController {
...
+   * Informs the correct listener that a direct message has been sent
+   * @param directMessageId
+   * @returns the listener that is recieving the message being sent
+   */
+  getDirectMessageListener(directMessageId: string): CoveyTownListener {
+    const matchingPlayerID = directMessageId.split(':')[1];
+    const directMessageListener = this._listeners.filter(
+      listener => listener.getAssociatedPlayer().id === matchingPlayerID,
+    );
+    return directMessageListener[0];
```
Comment (abigailhodge) [inline services/roomService/src/lib/CoveyTownController.ts]: Wait, shouldn't you be returning all of them? The way the frontend is set up, every player (including the one who sent the message) involved in a MessageChain needs to get the message.

**Review by abigailhodge — CHANGES_REQUESTED** [review]

File: services/roomService/src/client/CoveyTownsSocket.test.ts [hunk 5]
```diff
@@ -54,119 +60,283 @@ describe('TownServiceApiSocket', () => {
   });
   it('Rejects invalid CoveyTownIDs, even if otherwise valid session token', async () => {
     const town = await createTownForTesting();
-    const joinData = await apiClient.joinTown({coveyTownID: town.coveyTownID, userName: nanoid()});
-    const {socketDisconnected} = TestUtils.createSocketClient(server, joinData.coveySessionToken, nanoid());
+    const joinData = await apiClient.joinTown({
+      coveyTownID: town.coveyTownID,
```
Comment (MahemaS27) [inline services/roomService/src/client/CoveyTownsSocket.test.ts:64]: i dont know why this looks like i have a bunch of spaces between these lines when i open it in vscode its doesnt look like that

File: services/roomService/src/lib/CoveyTownController.ts [hunk 6]
```diff
@@ -195,4 +207,17 @@ export default class CoveyTownController {
       nearbyPlayers.includes(listener.getAssociatedPlayer()),
     );
   }
+
+  /**
+   * Informs the correct listener that a direct message has been sent
```
Comment (michelle-pine) [inline services/roomService/src/lib/CoveyTownController.ts]: i dont think this comment is fully accurate to what the function does. it just retrieves the listeners, it doesn't inform them

**Review by michelle-pine — APPROVED** [review]
LGTM!

**Review by abigailhodge — APPROVED** [review]
LGTM, thanks for the changes!

**Pull request merged**